scvi/data/anndata/_fields/_obsm_field.py

```diff
@@ -0,0 +1,103 @@
+import logging
+from typing import List, Optional
+
+import pandas as pd
+from anndata import AnnData
+
+from scvi.data.anndata import _constants
+
+from ._base_field import BaseAnnDataField
+
+logger = logging.getLogger(__name__)
+
+
+class BaseObsmField(BaseAnnDataField):
+    """An abstract AnnDataField for .obsm attributes in the AnnData data structure."""
+
+    _attr_name = _constants._ADATA_ATTRS.OBSM
+
+    def __init__(
+        self,
+        registry_key: str,
+        obs_keys: Optional[List[str]],
+    ) -> None:
+        super().__init__()
+        self._registry_key = registry_key
+        self._attr_key = f"_scvi_{registry_key}"
+        self._obs_keys = obs_keys if obs_keys is not None else []
+        self._is_empty = len(self.obs_keys) == 0
+
+    def _combine_obs_fields(self, adata: AnnData) -> None:
+        series = []
+        for key in self.obs_keys:
+            s = adata.obs[key]
+            series.append(s)
+        adata.obsm[self.attr_key] = pd.concat(series, axis=1)
+
+    @property
+    def registry_key(self):
+        return self._registry_key
+
+    @property
+    def obs_keys(self):
+        return self._obs_keys
+
+    @property
+    def attr_name(self):
+        return self._attr_name
+
+    @property
+    def attr_key(self):
+        return self._attr_key
+
+    @property
+    def is_empty(self) -> bool:
+        return self._is_empty
+
+
+class ContinuousObsmField(BaseObsmField):
```

Comment:
this should be able to cover e.g., protein expression for totalvi as well. Not sure this is the best name as counts are not continuous. Maybe `NonCategoricalObsmField` is better.

Also, in this PR, we could add the components to cover protein expression -- checking if counts, data reformatting to C contiguous memory, csr instead of coo etc.

Reply:
I think they are sufficiently different to make it a separate class that inherits from BaseObsmField class. This is because this class takes obs_keys and combines them into a new obsm field whereas the protein expression field is expected to be in a obsm field already, then reformatted etc. I tinkered around w/ making this class do both but it gets crowded